File: 198_house_robber/dp.py
class Solution:
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        """
        Dynamic Programming
        Complexity: O(n)
        Space: O(n)
        """
        if len(nums) == 0:
            return 0
        elif len(nums) == 1:
            return nums[0]
        
        max_money = [0]*len(nums)
        # boundary condition
        max_money[0] = nums[0]
        max_money[1] = max(max_money[0], nums[1])

        for i in range(2, len(nums)):
            max_money[i] = max(max_money[i-1], nums[i] + max_money[i-2])
            
        return max_money[-1]

File: 061_unique_path/dp.cpp
class Solution {
/*
 * Complexity: O(m*n)
 * Space: O(m*n)
 */
public:
    int uniquePaths(int m, int n) {
        vector< vector< unsigned long long> >dp(m, vector<unsigned long long>(n));
        for(int i = 0; i < m; ++i){
            for(int j = 0; j<n ; ++j){
                if(i == 0 && j == 0) dp[i][j] = 1;
                else if(i-1 >= 0 && j-1 >= 0)dp[i][j] = dp[i-1][j] + dp[i][j-1];
                else if (i-1 < 0) dp[i][j] = dp[i][j-1];
                else if (j-1 < 0) dp[i][j] = dp[i-1][j];
            }
        }
        return dp[m-1][n-1];
    }
};
